blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/credential/BlueOceanCredentialsProvider.java

```diff
@@ -0,0 +1,262 @@
+    }
+
+    private static <C extends Credentials> List<C> getCredentials(Class<C> type, AbstractFolder folder){
+        FolderPropertyImpl prop = (FolderPropertyImpl)folder.getProperties().get(FolderPropertyImpl.class);
+        if (prop != null) {
+            User user = User.get(prop.getUser(), false, Collections.emptyMap());
+            if(user != null){
+                Authentication auth = user.impersonate();
+                return CredentialsMatchers.filter(
+                        CredentialsProvider.lookupCredentials(
+                                type,
+                                Jenkins.getInstance(),
+                                auth,
+                                Collections.<DomainRequirement>emptyList()
+                        ),
+                        CredentialsMatchers.allOf(CredentialsMatchers.withId(prop.getId()),
+                                CredentialsMatchers.withScope(CredentialsScope.USER))
+                );
+            }
+        }
+        return Collections.emptyList();
+    }
+
+    @Nonnull
+    @Override
+    public <C extends IdCredentials> ListBoxModel getCredentialIds(@Nonnull Class<C> type,
+                                                                   @Nullable ItemGroup itemGroup,
+                                                                   @Nullable Authentication authentication,
+                                                                   @Nonnull List<DomainRequirement> domainRequirements,
+                                                                   @Nonnull CredentialsMatcher matcher) {
+        ListBoxModel result = new ListBoxModel();
+        if (itemGroup instanceof AbstractFolder) {
+            FolderPropertyImpl prop = (FolderPropertyImpl) ((AbstractFolder) itemGroup).getProperties().get(FolderPropertyImpl.class);
+            if (prop != null) {
+                result.add(DISPLAY_NAME, prop.getId());
+            }
+
+        }
+        return result;
+    }
+
+    @Override
+    @Nonnull
+    public String getDisplayName() {
+        return DISPLAY_NAME;
+    }
+
+    @Override
+    public CredentialsStore getStore(@CheckForNull ModelObject object) {
+        if(isApplicable(object)) {
+            return new FolderPropertyImpl.StoreImpl((AbstractFolder) object);
+        }
+        return null;
+    }
+
+    @Override
+    public Set<CredentialsScope> getScopes(ModelObject object) {
+        if(isApplicable(object)){
+            return ImmutableSet.of(CredentialsScope.USER);
+        }
+        return Collections.emptySet();
+    }
+
+    private boolean isApplicable(ModelObject object){
+        return object instanceof AbstractFolder &&
+                ((AbstractFolder)object).getProperties().get(FolderPropertyImpl.class) != null;
+    }
+
+    public static class FolderPropertyImpl extends AbstractFolderProperty<AbstractFolder<TopLevelItem>> {
+        private final String user;
+        private final String id;
```

Comment:
You don't necessarily need both. I can do MBP job creation only using my ssh key. https://gist.github.com/vivek/7a200b9cd6d645a78402b0ce5b1e56a8.

Our UI flow, 
- For Git case, we let user create either api key or ssh keys. In case of git, its ssh, or access key (username/password) but not both. 
- For GitHub, no ssh key, only GitHub accessKey.